README
NAME
    App::cpanlistchanges - list changes for CPAN modules

AUTHOR
    Tatsuhiko Miyagawa <miyagawa@bulknews.net>

    Tokuhiro Matsuno originally wrote the snippet to fetch Changes and
    compare with Algorithm::Diff if I remember correctly.

COPYRIGHT
    Copyright 2010- Tatsuhiko Miyagawa

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    cpan-listchanges



Changes
Revision history for Perl extension App::cpanlistchanges

{{$NEXT}}

0.08  2018-08-14 08:36:29 PDT
        - Use metacpan to fetch Changes (#6)

0.07  2015-07-21 01:37:05 PDT
        - update cpanmetadb URL

0.06  2013-04-07 01:29:17 PDT
        - Convert Milla and it fixes bad NAME in Makefile

0.05  Thu Jan 27 14:40:28 PST 2011
        - Support distributions released under multiple CPAN IDs
        - Automatically enable PAGER pipe. Use --no-pager to disable it
        - Fixed dependencies

0.04  Tue Jan 25 21:56:57 PST 2011
        - Fixed a subtle bug in the version output in the changes view

0.03  Tue Jan 25 21:47:46 PST 2011
        - Support 'HEAD' in the version comparison

0.02  Tue Jan 25 21:44:42 PST 2011
        - Added -h and pod2usage
        - Added --all which displays all the changelog regardless of local/remote versions
        - Added Module@from..to version comparison notation

0.01  Thu Dec 23 14:13:13 2010
        - original version


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "list changes for CPAN modules",
  "dist_author" => [
    "Tatsuhiko Miyagawa <miyagawa\@bulknews.net>"
  ],
  "dist_name" => "cpan-listchanges",
  "dist_version" => "0.06",
  "license" => "perl",
  "module_name" => "cpan::listchanges",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Algorithm::Diff" => 0,
    "CPAN::DistnameInfo" => 0,
    "LWP" => 0,
    "Module::Metadata" => 0,
    "Pod::Usage" => 0,
    "Try::Tiny" => 0,
    "YAML" => 0,
    "perl" => "5.008001"
  },
  "script_files" => [
    "script/cpan-listchanges"
  ],
  "test_requires" => {
    "Test::More" => "0.88",
    "Test::Requires" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/App/cpanlistchanges.pm
package App::cpanlistchanges;

use strict;
use 5.008_001;
our $VERSION = '0.07';

use Algorithm::Diff;
use CPAN::DistnameInfo;
use Getopt::Long;
use IO::Handle;
use Module::Metadata;
use LWP::UserAgent;
use YAML;
use Try::Tiny;
use Pod::Usage;
use version;

sub new {
    bless {
        all => 0,
        use_pager => 1,
    }, shift;
}

sub run {
    my($self, @args) = @_;

    Getopt::Long::GetOptionsFromArray(
        \@args,
        "all|a", \$self->{all},
        "help",  sub { Pod::Usage::pod2usage(0) },
        "pager!", \$self->{use_pager},
    );

    for my $mod (@args) {
        $self->show_changes($mod);
    }
}

sub get {
    my $self = shift;

    $self->{ua} ||= do{
        my $ua = LWP::UserAgent->new(agent => "cpan-listchanges/$VERSION");
        $ua->env_proxy;
        $ua;
    };

    $self->{ua}->get(@_)->content;
}

sub show_changes {
    my($self, $mod) = @_;

    my($from, $to);
    if ($mod =~ s/\@\{?(.+)\}?$//) {
        ($from, $to) = split /\.\./, $1;
        $to = undef if $to eq 'HEAD';
    }

    my $dist = try { YAML::Load( $self->get("http://cpanmetadb.plackperl.org/v1.0/package/$mod") ) };
    unless ($dist->{distfile}) {
        warn "Couldn't find a module '$mod'. Skipping.\n";
        return;
    }

    my $meta = Module::Metadata->new_from_module($mod);
    my $info = CPAN::DistnameInfo->new($dist->{distfile});

    $from ||= $meta->{version};
    $to   ||= $info->{version};

    unless ($self->{all} or $from) {
        warn "You don't have the module '$mod' installed locally. Skipping.\n";
        return;
    }

    unless ($self->{all} or $to) {
        warn "Couldn't find the module '$mod' on CPAN MetaDB. Skipping.\n";
        return;
    }

    if (!$self->{all} and version->new($from) >= version->new($to)) {
        warn "You have the latest version of $info->{dist} ($to). Skipping.\n";
        return;
    }

    my $get_changes = sub {
        my $version = shift;
        return $self->get(
            "https://fastapi.metacpan.org/source/$info->{cpanid}/$info->{dist}-$version/Changes"
        );
    };

    my $new_changes = $get_changes->($to);
    unless ($new_changes) {
        warn "Can't find Changes for $info->{dist}-$to. Skipping.\n";
        return;
    }

    if ($self->{all}) {
        $self->print("=== Changes for $info->{dist}\n\n$new_changes\n");
        return;
    }

    my $old_changes = $get_changes->($from) || '';

    my $diff = Algorithm::Diff->new(
        [ split /\n/, $old_changes ],
        [ split /\n/, $new_changes ],
    );
    $diff->Base(1);

    my $result;
    while ($diff->Next()) {
        next if $diff->Same();
        $result .= "$_\n" for $diff->Items(2);
    }

    if ($result) {
        $self->print("=== Changes between $from and $to for $info->{dist}\n\n$result\n");
    } else {
        warn "Couldn't find changes between $from and $to for $info->{dist}\n";
    }
}

sub print {
    my $self = shift;
    $self->{pager} ||= ($self->{use_pager} and $ENV{PAGER})
        ? do {
            open my $io, "| $ENV{PAGER}" or die $!;
            $io;
        } : \*STDOUT;

    $self->{pager}->print(@_);
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

App::cpanlistchanges - list changes for CPAN modules

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

Tokuhiro Matsuno originally wrote the snippet to fetch Changes and
compare with Algorithm::Diff if I remember correctly.

=head1 COPYRIGHT

Copyright 2010- Tatsuhiko Miyagawa

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<cpan-listchanges>

=cut


Makefile.PL

use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "list changes for CPAN modules",
  "AUTHOR" => "Tatsuhiko Miyagawa <miyagawa\@bulknews.net>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "cpan-listchanges",
  "EXE_FILES" => [
    "script/cpan-listchanges"
  ],
  "LICENSE" => "perl",
  "NAME" => "cpan::listchanges",
  "PREREQ_PM" => {
    "Algorithm::Diff" => 0,
    "CPAN::DistnameInfo" => 0,
    "LWP" => 0,
    "Module::Metadata" => 0,
    "Pod::Usage" => 0,
    "Try::Tiny" => 0,
    "YAML" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.88",
    "Test::Requires" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



